src/server/game/Entities/Object/Object.cpp

```diff
@@ -2320,52 +2396,70 @@ void WorldObject::MovePositionToFirstCollision(Position &pos, float dist, float
         return;
     }
 
-    UpdateAllowedPositionZ(destx, desty, destz);
-    bool col = VMAP::VMapFactory::createOrGetVMapManager()->getObjectHitPos(GetMapId(), pos.m_positionX, pos.m_positionY, pos.m_positionZ, destx, desty, destz, destx, desty, destz, -0.5f);
-
-    // collision occured
+    bool const col = ComputeCollisionPosition(pos, { destx, desty, destz }, destx, desty, destz);
     if (col)
-    {
-        // move back a bit
-        destx -= CONTACT_DISTANCE * std::cos(angle);
-        desty -= CONTACT_DISTANCE * std::sin(angle);
         dist = std::sqrt((pos.m_positionX - destx)*(pos.m_positionX - destx) + (pos.m_positionY - desty)*(pos.m_positionY - desty));
-    }
 
-    // check dynamic collision
-    col = GetMap()->getObjectHitPos(GetPhaseMask(), pos.m_positionX, pos.m_positionY, pos.m_positionZ, destx, desty, destz, destx, desty, destz, -0.5f);
+    destz = SelectBestZForDestination(destx, desty, destz, col);
```

Comment:
> // You should check for collision again after this one has been called.

Reply:
Yes, still looking into this. I didn't like how I did it before as we might had to reselect a better z until no more collision but that seemed to much for me. Also got infinity loops while doing it lol.
So "ignoring" that comment makes the behavior a bit more like what happened previously

Reply:
I'm going to re add it anyway as it should fix more than break things (no more blinking of the roof in WSG!), with a comment tho